pkg/blobmanager/s3/backend.go

```diff
@@ -100,29 +149,41 @@ func (b *Backend) Exists(ctx context.Context, digest string) (bool, error) {
 
 func (b *Backend) Upload(ctx context.Context, r io.Reader, resource *pb.CASResource) error {
 	uploader := s3manager.NewUploaderWithClient(b.client)
-
-	_, err := uploader.UploadWithContext(ctx, &s3manager.UploadInput{
+	input := &s3manager.UploadInput{
 		Bucket: aws.String(b.bucket),
 		Key:    aws.String(resourceName(resource.Digest)),
 		Body:   r,
-		// Check that the object is uploaded correctly
-		ChecksumSHA256: aws.String(hexSha256ToBinaryB64(resource.Digest)),
 		Metadata: map[string]*string{
 			annotationNameAuthor:   aws.String(backend.AuthorAnnotation),
 			annotationNameFilename: aws.String(resource.FileName),
 		},
-	})
+	}
 
-	return err
+	if b.checksumVerificationEnabled() {
+		// Check that the object is uploaded correctly
+		input.ChecksumSHA256 = aws.String(hexSha256ToBinaryB64(resource.Digest))
+	}
+
+	if _, err := uploader.UploadWithContext(ctx, input); err != nil {
```

Comment:
on another topic, does this do streaming already?

Reply:
The underlying library does, the `service` layer that servers bytestream does not.